### PR TITLE
Review Blog for Publication

### PR DESCRIPTION
**Pull request by nraboy**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -0,0 +1,335 @@
...
+	"go.mongodb.org/mongo-driver/mongo"
+	"go.mongodb.org/mongo-driver/mongo/options"
+)
+
+type Connection struct {
+	database *mongo.Database
+}
+
+func main() {
+	client, err := mongo.NewClient(options.Client().ApplyURI(os.Getenv("ATLAS_URI")))
```
Comment (divjotarora) [inline README.md]: Use `mongo.Connect` here for a more concise way of creating a client.

File: README.md [hunk 2]
```diff
@@ -0,0 +1,335 @@
...
+	if err != nil {
+		log.Fatal(err)
```
Comment (divjotarora) [inline README.md]: Something I didn't notice in earlier reviews: `log.Fatal` actually runs `os.Exit`, so any registered `defer` statements do not run. If possible, replace uses of `log.Fatal` with `panic`.

File: README.md [hunk 3]
```diff
@@ -0,0 +1,335 @@
...
+	}
+	ctx, _ := context.WithTimeout(context.Background(), 10*time.Second)
```
Comment (divjotarora) [inline README.md]: We're actually trying to deprecate the `Context` parameter for the `Connect` function because function doesn't do any blocking I/O, so the timeout is meaningless. Change to `context.Background()` or `context.TODO()` and remove the text below about the 10 second timeout.

File: README.md [hunk 4]
```diff
@@ -0,0 +1,335 @@
...
+	database := client.Database("alexa")
+	connection := Connection{
+		database: database,
+	}
+}
+```
+
+In the `main` function we are creating a client using the connection string of our cluster. In this case, I'm using an environment variable on my computer that points to my MongoDB Atlas cluster. Feel free to configure that connection string however you feel the most confident.
+
+Before attempting to connect, we are setting a generous timeout of ten seconds. Upon connecting, we are getting a handle of an `alexa` database and storing it in a `Connection` data structure. If the `alexa` database doesn't exist, it will create one if we try to create data.
```
Comment (divjotarora) [inline README.md]: `If the `alexa` database doesn't exist, it will create one if we try to create data.`

To me, this implies that the `Client.Database` call creates a new database object in MongoDB, but it actually doesn't do any I/O. It's creating a Go representation of a MongoDB database, and the first write operation will create the database in the server if necessary.

File: README.md [hunk 5]
```diff
@@ -0,0 +1,335 @@
...
+}
+```
+
+There is no doubt that our documents could be more extravagant, but for this example it will work out fine. Within the MongoDB Atlas cluster, create the **alexa** database if it doesn't already exist and add a document modeled like the above in a **recipes** collection.
+
+In the `main.go` file of the project, add the following data structure:
+
+```go
+type Recipe struct {
+	Id          primitive.ObjectID `bson:"_id"`
```
Comment (divjotarora) [inline README.md]: This should be named `ID`.

File: README.md [hunk 6]
```diff
@@ -0,0 +1,335 @@
...
+
+```go
+case "GetIngredientsForRecipeIntent":
+    var recipe Recipe
+    recipesCollection := connection.database.Collection("recipes")
+    recipeName := request.Body.Intent.Slots["recipe"].Value
+    if recipeName == "" {
+        return alexa.Response{}, errors.New("Recipe name is not present in the request")
+    }
+    if err := recipesCollection.FindOne(ctx, bson.M{"name": recipeName}).Decode(&recipe); err != nil {
```
Comment (divjotarora) [inline README.md]: How do errors behave for Alexa skills? For example, if a recipe isn't found, `FindOne` will return `mongo.ErrNoDocuments`. Should that be reported as an error here or as a `alexa.NewSimpleResponse`?

File: README.md [hunk 7]
```diff
@@ -0,0 +1,335 @@
...
+
+Now let's take a look at the `GetRecipeFromIngredientsIntent` intent logic:
+
+```go
+case "GetRecipeFromIngredientsIntent":
+    var recipes []Recipe
+    recipesCollection := connection.database.Collection("recipes")
+    ingredient1 := request.Body.Intent.Slots["ingredientone"].Value
+    ingredient2 := request.Body.Intent.Slots["ingredienttwo"].Value
+    cursor, err := recipesCollection.Find(ctx, bson.M{"ingredients": bson.D{{"$all", bson.A{ingredient1, ingredient2}}}})
```
Comment (divjotarora) [inline README.md]: Might be more readable to split the `bson.M` across multiple lines:
```
bson.M{
    "ingredients": bson.D{
        {"$all", bson.A{ingredient1, ingredient2},
    },
}
```

File: README.md [hunk 8]
```diff
@@ -0,0 +1,335 @@
...
+        return alexa.Response{}, err
+    }
+    if err = cursor.All(ctx, &recipes); err != nil {
+        return alexa.Response{}, err
+    }
+    recipeList := ""
+    for _, recipe := range recipes {
+        recipeList += recipe.Name
+    }
+    response = alexa.NewSimpleResponse("Recipes", recipeList)
```
Comment (divjotarora) [inline README.md]: `recipeList` will be one string with no delimiters but the `GetIngredientsForRecipeIntent` intent returns one comma-delimited string created via `strings.Join`. Are both correct? Even if they are both correct, maybe we should stick to one method for consistency?

File: main.go [hunk 9]
```diff
@@ -0,0 +1,85 @@
+package main
```
Comment (divjotarora) [inline main.go:1]: The code needs two improvements:

1. All exported types should have a documentation comment beginning with the type name (e.g. `Connection stores a handle to a MongoDB database`). This is to make basic linters pass without complaints.

2. More whitespace is needed for readability. Every logical block of code should be separated by a new line. Let me know if you need suggestions about where to add whitespace.

File: README.md [hunk 10]
```diff
@@ -0,0 +1,335 @@
...
+	"time"
+
+	"github.com/arienmalec/alexa-go"
+	"github.com/aws/aws-lambda-go/lambda"
+	"go.mongodb.org/mongo-driver/mongo"
+	"go.mongodb.org/mongo-driver/mongo/options"
+)
+
+type Connection struct {
+	database *mongo.Database
```
Comment (divjotarora) [inline README.md]: Instead of storing a handle to a `Database` object that's only used to get handles to `Collection` objects, why not store a handle to the `recipes` collection instead?

File: README.md (hunk 3 above)
Comment (nraboy) [reply README.md]: The documentation is still suggesting the use of `Context` for connecting:

https://github.com/mongodb/mongo-go-driver#usage

Can we have this updated for consistency?

Comment (divjotarora) [reply README.md]: Yeah, that should be updated. We'll update it when we do the ticket to deprecate the parameter.

File: README.md (hunk 6 above)
Comment (nraboy) [reply README.md]: The Lambda SDK should properly process the error.

**Comment by nraboy** [comment]
Please give it another pass.

Thanks,

**Review by divjotarora — APPROVED** [review]
LGTM

File: README.md [hunk 11]
```diff
@@ -0,0 +1,331 @@
+# Developing Alexa Skills with MongoDB and Golang
+
+The popularity of Amazon Alexa and virtual assistants in general is no question, huge. Having a web application and mobile application isn't enough anymore, and now you need to start supporting voice operated applications.
```
Comment (ljhaywar) [inline README.md]: link to Alexa?

Comment (ljhaywar) [inline README.md]: enough for who?

Comment (ljhaywar) [inline README.md]: everyone needs to start supporting voice apps?

File: README.md [hunk 12]
```diff
@@ -0,0 +1,331 @@
...
+
+So what does it take to create something for Alexa? How different is it from creating a web application?
+
+In this tutorial we're going to see how to create an Amazon Alexa Skill that interacts with a MongoDB cluster using the Go programming language (Golang) and AWS Lambda.
```
Comment (ljhaywar) [inline README.md]: Many people may not know what a skill is. Can you define it and/or provide a link?

File: README.md [hunk 13]
```diff
@@ -0,0 +1,331 @@
...
+
+## The Requirements
+
+There are a few requirements that must be met prior to starting this tutorial:
```
Comment (ljhaywar) [inline README.md]: What is "there"?  Sentences that start with "there" can typically be made stronger by restructuring to have a clear subject.

File: README.md [hunk 14]
```diff
@@ -0,0 +1,331 @@
...
+
+- Golang must be installed and configured
+- A MongoDB Atlas cluster
```
Comment (ljhaywar) [inline README.md:16]: first is a verb phrase. second is not.

File: README.md [hunk 15]
```diff
@@ -0,0 +1,331 @@
...
+
+If you don't have a [MongoDB Atlas](https://www.mongodb.com/cloud) cluster, you can configure one for free. For this example an M0 cluster is more than sufficient.
```
Comment (ljhaywar) [inline README.md:18]: do you want to use your bitly atlas link?

File: README.md [hunk 16]
```diff
@@ -0,0 +1,331 @@
...
+
+> Using promotional code NRABOY200 will give you $200 in premium credit on your Atlas account for a more powerful cluster beyond the forever free M0 cluster.
+
+Having an Amazon Echo or other Amazon Alexa enabled device is not necessary to be successful with this tutorial. Amazon offers a really great simulator that can be used directly in the web browser.
+
+## Designing an Alexa Skill with an Invocation Term and Sample Utterances
+
+When it comes to building an Alexa Skill, it doesn't matter if you start with the code or the design. For this tutorial we're going to start with the design, directly in the [Amazon Developer Portal](https://developer.amazon.com) for Alexa.
```
Comment (ljhaywar) [inline README.md:28]: Same thing as the sentence above that started with "there."  What is "it?" Can you rephrase to have a clear subject?

File: README.md (hunk 14 above)
Comment (ljhaywar) [inline README.md:16]: Looks like you also need an amazon developer account

File: README.md [hunk 17]
```diff
@@ -0,0 +1,331 @@
...
+
+Sign into the portal and choose to create a new custom Skill. After creating the Skill, you'll be brought to a dashboard with several checklist items:
```
Comment (ljhaywar) [inline README.md:30]: Do you have a suggested name?

Comment (ljhaywar) [inline README.md:30]: After creating a custom skill, I was asked if I wanted to start from scratch, fact skill, quiz game skill, or high-low game skill

File: README.md [hunk 18]
```diff
@@ -0,0 +1,331 @@
...
+
+![Amazon Alexa Portal](amazon-alexa-portal.png "Amazon Alexa Portal")
+
+In the checklist, you should take note of the following:
+
+- Invocation Name
+- Intents, Samples, and Slots
+- Endpoint
+
+There are other items, one being optional and the other being checked naturally as the others complete.
```
Comment (ljhaywar) [inline README.md:40]: What is "there"

File: README.md [hunk 19]
```diff
@@ -0,0 +1,331 @@
...
+
+For our invocation name, let's use **recipe manager**, something that is easy to remember and easy to pronounce. With the invocation name in place, we can anticipate using our Skill like the following:
+
+```plaintext
+Alexa, ask Recipe Manager to INTENT
+```
+
+The user would not literally speak **INTENT** in the command. The intent is the command that will be defined through sample utterances, also known as sample phrases or data. You can, and probably should, have multiple intents for your Skill.
+
+Let's start by creating an intent titled **GetIngredientsForRecipeIntent** with the following sample utterances:
```
Comment (ljhaywar) [inline README.md:52]: After I input my skill invocation name, I wasn't sure what to do.  Should I click Save Model?  What do I need to do to get to the page where I create the intent?

Comment (ljhaywar) [inline README.md:52]: should this be a custom intent?

File: README.md [hunk 20]
```diff
@@ -0,0 +1,331 @@
...
+
+```plaintext
+what ingredients do i need for {recipe}
```
Comment (ljhaywar) [inline README.md:55]: when I pasted this in, a dialog popped up asking me about creating a new slot

File: README.md [hunk 21]
```diff
@@ -0,0 +1,331 @@
...
+what do i need to cook {recipe}
+to cook {recipe} what ingredients do i need
+```
+
+There are a few things to note about the above phrases:
+
+- The `{recipe}` tag is a slot variable which is going to be user defined when spoken.
+- Every possible spoken phrase to execute the command should be listed.
+
+Alexa operates from machine learning, so the more sample data the better. When defining the `{recipe}` variable, it should be assigned a type of `AMAZON.Food`.
```
Comment (ljhaywar) [inline README.md:65]: I needed to know this earlier

File: README.md [hunk 22]
```diff
@@ -0,0 +1,331 @@
...
+
+When all said and done, you could execute the intent by doing something like:
+
+```plaintext
+Alexa, ask Recipe Manager what do I need to cook Chocolate Chip Cookies
+```
+
+Having one intent in your Alexa Skill is no fun, so let's create another intent with its own set of sample phrases. Choose to create a new intent titled `GetRecipeFromIngredientsIntent` with the following sample utterances:
```
Comment (ljhaywar) [inline README.md:73]: Custom intent?

File: README.md [hunk 23]
```diff
@@ -0,0 +1,331 @@
...
+
+```plaintext
+what can i cook with {ingredientone} and {ingredienttwo}
```
Comment (ljhaywar) [inline README.md:76]: Would be helpful to know any details about ingredientone and ingredienttwo here as this is when I'm being asked to create them

File: README.md [hunk 24]
```diff
@@ -0,0 +1,331 @@
...
+what are some recipes with {ingredientone} and {ingredienttwo}
+if i have {ingredientone} and {ingredienttwo} what can i cook
+```
+
+This time around we're using two slot variables instead of one. Like previously mentioned, it is probably a good idea to add significantly more sample utterances to get the best results. Alexa needs to be able to process the data to send to your Lambda function.
```
Comment (ljhaywar) [inline README.md:81]: what is "it"?

File: README.md [hunk 25]
```diff
@@ -0,0 +1,331 @@
...
+
+At this point in time, the configuration in the Alexa Developer Portal is about complete. The exception being the endpoint which doesn't exist yet.
```
Comment (ljhaywar) [inline README.md:83]: which phrases should have a comma before them

File: README.md [hunk 26]
```diff
@@ -0,0 +1,331 @@
...
+
+## Building a Lambda Function with Golang and MongoDB
+
+Alexa, for the most part should be able to direct requests, so now we need to create our backend to receive and process them. This is where Lambda, Go, and MongoDB come into play.
```
Comment (ljhaywar) [inline README.md:87]: add a comma after part

File: README.md [hunk 27]
```diff
@@ -0,0 +1,331 @@
...
+package main
+
+func main() { }
+```
+
+With the boilerplate code added, now we can install the MongoDB Go driver. To do this, you could in theory do a `go get`, but the preferred approach as of now is to use the [dep](https://github.com/golang/dep) package management tool for Golang. To do this, after having installed the tool, execute the following:
+
+```bash
+dep init
+dep ensure -add "go.mongodb.org/mongo-driver/mongo"
```
Comment (ljhaywar) [inline README.md:101]: I received the following warning:

Fetching sources...

```
"go.mongodb.org/mongo-driver/mongo" is not imported by your project, and has been temporarily added to Gopkg.lock and vendor/.
If you run "dep ensure" again before actually importing it, it will disappear from Gopkg.lock and vendor/.
```

File: README.md [hunk 28]
```diff
@@ -0,0 +1,331 @@
...
+```
+
+We're using `dep` so that way the version of the driver that we're using in our project is version locked.
+
+In addition to the MongoDB Go driver, we're also going to need to get the [AWS Lambda SDK for Go](https://github.com/aws/aws-lambda-go) as well as an [unofficial SDK for Alexa](https://github.com/arienmalec/alexa-go), since no official SDK exists. To do this, we can execute:
+
+```bash
+dep ensure -add "github.com/arienmalec/alexa-go"
+dep ensure -add "github.com/aws/aws-lambda-go/lambda"
```
Comment (ljhaywar) [inline README.md:110]: Received warnings:

```
Warning: the following project(s) have [[constraint]] stanzas in Gopkg.toml:

  ✗  github.com/arienmalec/alexa-go
  ✗  go.mongodb.org/mongo-driver

However, these projects are not direct dependencies of the current project:
they are not imported in any .go files, nor are they in the 'required' list in
Gopkg.toml. Dep only applies [[constraint]] rules to direct dependencies, so
these rules will have no effect.

Either import/require packages from these projects so that they become direct
dependencies, or convert each [[constraint]] to an [[override]] to enforce rules
on these projects, if they happen to be transitive dependencies.

Fetching sources...

"github.com/aws/aws-lambda-go/lambda" is not imported by your project, and has been temporarily added to Gopkg.lock and vendor/.
If you run "dep ensure" again before actually importing it, it will disappear from Gopkg.lock and vendor/.
```

File: README.md [hunk 29]
```diff
@@ -0,0 +1,331 @@
...
+
+	defer client.Disconnect(ctx)
+
+	connection := Connection{
+		collection: client.Database("alexa").Collection("recipes"),
+	}
+}
+```
+
+In the `main` function we are creating a client using the connection string of our cluster. In this case, I'm using an environment variable on my computer that points to my MongoDB Atlas cluster. Feel free to configure that connection string however you feel the most confident.
```
Comment (ljhaywar) [inline README.md:146]: Do you explain this more in another post?  If so, can you point us to that?  People might be wondering where do I get the Atlas URI?  I was wondering, how do I set an os env variable?

File: README.md [hunk 30]
```diff
@@ -0,0 +1,331 @@
...
+
+Upon connecting, we are getting a handle of a `recipes` collection for an `alexa` database and storing it in a `Connection` data structure. Because we won't be writing any data in this example, both the `alexa` database and the `recipes` collection should exist prior to running this application.
+
+You can check out more information about connecting to MongoDB with the Go programming language in a [previous tutorial](https://www.mongodb.com/blog/post/quick-start-golang--mongodb--starting-and-setup) I wrote.
```
Comment (ljhaywar) [inline README.md:150]: Oh, I see this now.  Maybe this should be moved up.

File: README.md [hunk 31]
```diff
@@ -0,0 +1,331 @@
...
+
+So why are we storing the collection handle in a `Connection` data structure?
+
+AWS Lambda behaves a little differently when it comes to web applications. Instead of running the `main` function and then remaining alive for as long as your server remains alive, Lambda functions tend to suspend or shutdown when they are not used. For this reason, we cannot rely on our connection being available and we also don't want to establish too many connections to our database in the scenario where our function hasn't shut down. To handle this, we can pass the connection from our `main` function to our logic function.
```
Comment (ljhaywar) [inline README.md:154]: I don't think you've explicitly told us yet what our lamdba function is doing or that it will be our logic function

File: README.md [hunk 32]
```diff
@@ -0,0 +1,331 @@
...
+
+	connection := Connection{
+		collection: client.Database("alexa").Collection("recipes"),
+	}
+
+	lambda.Start(connection.IntentDispatcher)
+}
+```
+
+Notice in the above code that we've added a `lambda.Start` call in our `main` function that points to an `IntentDispatcher` function. We're designing this function to use the connection information established in the `main` function, which based on our Lambda knowledge, may not run every time the function is executed.
```
Comment (ljhaywar) [inline README.md:197]: Is "this function" the main function?

File: README.md [hunk 33]
```diff
@@ -0,0 +1,331 @@
...
+
+So we've got the foundation to our Alexa Skill in place. Now we need to design the logic for each of our intents that were previously defined in the Alexa Developer Portal.
+
+Since this is going to be a recipe related Skill, let's model our MongoDB documents like the following:
+
+```json
```
Comment (ljhaywar) [inline README.md:203]: I tried pasting this in to the Atlas ui, and it gave me an error until I removed _id

File: README.md [hunk 34]
```diff
@@ -0,0 +1,331 @@
...
+	default:
+		response = alexa.NewSimpleResponse("Unknown Request", "The intent was unrecognized")
+	}
+	return response, nil
+}
+```
+
+Remember the two intents from the Alexa Developer Portal? We need to assign logic to them.
+
+Essentially, we're going to do some database logic and then use the `NewSimpleResponse` function to create a response the the results.
```
Comment (ljhaywar) [inline README.md:248]: "do some database logic" - this wording feels off to me

Comment (ljhaywar) [inline README.md:248]: "response **the the** results"

File: README.md [hunk 35]
```diff
@@ -0,0 +1,331 @@
...
+	if recipeName == "" {
+		return alexa.Response{}, errors.New("Recipe name is not present in the request")
+	}
+	if err := connection.collection.FindOne(ctx, bson.M{"name": recipeName}).Decode(&recipe); err != nil {
+		return alexa.Response{}, err
+	}
+	response = alexa.NewSimpleResponse("Ingredients", strings.Join(recipe.Ingredients, ", "))
+```
+
+In the above snippet, we are getting the slot variable that was passed and are issuing a `FindOne` query against the collection. The filter for the query says that the `name` field of the document must match the recipe that was passed in as a slot variable.
```
Comment (ljhaywar) [inline README.md:265]: Is there any sort of fuzzy search?  Could you add full text search support to this?  Maybe that's something to discuss at the end of the article

File: README.md [hunk 36]
```diff
@@ -0,0 +1,331 @@
...
+		return alexa.Response{}, err
+	}
+	var recipeList []string
+	for _, recipe := range recipes {
+		recipeList = append(recipeList, recipe.Name)
+	}
+	response = alexa.NewSimpleResponse("Recipes", strings.Join(recipeList, ", "))
+```
+
+In the above snippet, we are taking both slot variables that represent ingredients and are using them in a `Find` query on the collection. This time around we are using the `$all` operator because we want to filter for all recipes that contain both ingredients anywhere in the array.
```
Comment (ljhaywar) [inline README.md:294]: Let's say I have recipes for toast with jam as well as peanut butter and jelly sandwiches.  If I ask Alexa to find me a recipe for the ingredients bread, peanut butter, and jam, I would expect to only get the recipe for peanut butter and jelly.  If I ask Alexa to find me a recipe for the ingredients bread and jam, I would expect to get both recipes.  I think what you're describing is the opposite of that.

File: README.md [hunk 37]
```diff
@@ -0,0 +1,331 @@
...
+
+So what's happening in the above commands? First we are building a Linux compatible binary. We're doing this because if you're developing on Mac or Windows, you're going to end up with a binary that is incompatible. By defining the operating system, we're telling Go what to build for.
+
+For more information on cross-compiling with Go, check out my [previous tutorial](https://www.thepolyglotdeveloper.com/2017/04/cross-compiling-golang-applications-raspberry-pi/) on the subject.
+
+Next, we are creating an archive of our binary. It is important to replace the `project-name` with that of your actual binary name. It is important to remember the name of the file as it is used in the Lambda dashboard.
+
+![Amazon Lambda Dashboard](amazon-lambda-portal.jpg "Amazon Lambda Dashboard")
+
+When you choose to create a new Lambda function within AWS, make sure Go is the development technology. Choose to upload the ZIP file and add the name of the binary as the handler.
```
Comment (ljhaywar) [inline README.md:319]: Where do I go to create the function?

File: README.md [hunk 38]
```diff
@@ -0,0 +1,331 @@
...
+dep init
+dep ensure -add "go.mongodb.org/mongo-driver/mongo"
+```
+
+We're using `dep` so that way the version of the driver that we're using in our project is version locked.
+
+In addition to the MongoDB Go driver, we're also going to need to get the [AWS Lambda SDK for Go](https://github.com/aws/aws-lambda-go) as well as an [unofficial SDK for Alexa](https://github.com/arienmalec/alexa-go), since no official SDK exists. To do this, we can execute:
+
+```bash
+dep ensure -add "github.com/arienmalec/alexa-go"
```
Comment (ljhaywar) [inline README.md:109]: I chatted with you about the errors I hit with these imports

File: README.md [hunk 39]
```diff
@@ -0,0 +1,331 @@
...
+
+Next, we are creating an archive of our binary. It is important to replace the `project-name` with that of your actual binary name. It is important to remember the name of the file as it is used in the Lambda dashboard.
+
+![Amazon Lambda Dashboard](amazon-lambda-portal.jpg "Amazon Lambda Dashboard")
+
+When you choose to create a new Lambda function within AWS, make sure Go is the development technology. Choose to upload the ZIP file and add the name of the binary as the handler.
+
+Now it comes down to linking Alexa with Lambda.
+
+Take note of the **ARN** value of your Lambda function. This will be added in the Alexa Portal. Also, make sure you add the Alexa Skills Kit as a trigger to the function. It is as simple as selecting it from the list.
```
Comment (ljhaywar) [inline README.md:323]: Do I need to enable skill id verification? If so, what values should I be using?

File: README.md [hunk 40]
```diff
@@ -0,0 +1,331 @@
...
+
+Navigate back to the Alexa Developer Portal and choose the **Endpoint** checklist item. Add the ARN value and choose to build the Skill.
```
Comment (ljhaywar) [inline README.md]: a link to the portal would be helpful here

Comment (ljhaywar) [inline README.md]: I was a little confused about where to paste the value.  I think "Default Region" is the right spot

Comment (ljhaywar) [inline README.md]: I don't see a build the Skill option on this page.  I saved the endpoint.  Then I poked around to try to figure out where to build the skill.  I think what I want to do is Build Model, but I have a bunch of errors related to my slots.

File: README.md (hunk 21 above)
Comment (ljhaywar) [reply README.md:65]: I hit build errors because I didn't set the slot types for ingredientone and ingredienttwo to amazon.food

File: README.md [hunk 41]
```diff
@@ -0,0 +1,331 @@
...
+
+When the Skill is done building, you can test it using the simulator that Amazon offers as part of the Alexa Developer Portal.
+
+## Conclusion
+
+You just saw how to build an Alexa Skill with [MongoDB](https://www.mongodb.com), Golang, and AWS Lambda. It is great to know how to develop applications for voice assistants like Alexa because they are becoming increasingly popular, and the good news is that they aren't any more difficult than writing standard applications.
```
Comment (ljhaywar) [inline README.md]: What is "it"?

File: README.md [hunk 42]
```diff
@@ -0,0 +1,331 @@
...
+
+When you choose to create a new Lambda function within AWS, make sure Go is the development technology. Choose to upload the ZIP file and add the name of the binary as the handler.
+
+Now it comes down to linking Alexa with Lambda.
+
+Take note of the **ARN** value of your Lambda function. This will be added in the Alexa Portal. Also, make sure you add the Alexa Skills Kit as a trigger to the function. It is as simple as selecting it from the list.
+
+Navigate back to the Alexa Developer Portal and choose the **Endpoint** checklist item. Add the ARN value and choose to build the Skill.
+
+When the Skill is done building, you can test it using the simulator that Amazon offers as part of the Alexa Developer Portal.
```
Comment (ljhaywar) [inline README.md]: I could use a few steps here to explain how to test this out

**Review by ljhaywar — CHANGES_REQUESTED** [review]
I'm going to stop here for today.  I got stuck and I'm not sure how to debug.  I messaged you with a screenshot on Slack.

File: README.md (hunk 11 above)
Comment (nraboy) [reply README.md]: added "... isn't enough for most organizations anymore ..." so no longer implies everyone.
